### PR TITLE
RHEL matcher

### PR DESCRIPTION
**Pull request by hdonnay**
This series of commits creates a matcher that works from an OVAL RPM database and adds it to the default set.

Also contains bugfixes and utilities created along the way.

### CODE REVIEW

File: rpm/matcher.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	return true
+}
+
+func (*Matcher) Query() []driver.MatchExp {
+	return []driver.MatchExp{
+		driver.PackageDistributionCPE,
+	}
+}
+
+func (*Matcher) Vulnerable(record *claircore.IndexRecord, vuln *claircore.Vulnerability) bool {
```
Comment (ldelossa) [inline rpm/matcher.go]: I'd put a little comment here just explaining when/why we'd want out the cmp function.

File: internal/vulnstore/postgres/querybuilder.go [hunk 2]
```diff
@@ -24,58 +24,39 @@ func getBuilder(matchers []driver.MatchExp) (string, []driver.MatchExp, error) {
 	seen := make(map[driver.MatchExp]struct{})
 	deduped := []driver.MatchExp{}
 	for _, m := range matchers {
+		if _, ok := seen[m]; ok {
+			continue
+		}
+		var ex goqu.Ex
 		switch m {
+		case driver.PackageName: // ???
```
Comment (ldelossa) [inline internal/vulnstore/postgres/querybuilder.go]: Can we just nix these from this switch and from MatchExps ?

Comment (hdonnay) [reply internal/vulnstore/postgres/querybuilder.go]: Yeah, I think removing them and renaming those fields is in order. Is that a separate PR?

Comment (ldelossa) [reply internal/vulnstore/postgres/querybuilder.go]: @hdonnay you can make a judgement call on that. If it's a far-reaching rename and touches a lot maybe we punt to a dedicated PR. If it seems quick then by all means do it in this one.

File: rpm/matcher.go (hunk 1 above)
Comment (hdonnay) [reply rpm/matcher.go]: OK

File: internal/vulnstore/postgres/get.go [hunk 3]
```diff
@@ -11,20 +14,34 @@ import (
...
 	query, dedupedMatchers, err := getBuilder(opts.Matchers)
+	if err != nil {
+		return nil, err
+	}
+	h := fnv.New64a()
+	if _, err := io.WriteString(h, query); err != nil {
+		return nil, err
+	}
+	name := hex.EncodeToString(h.Sum(nil))
+	log.Debug().Str("name", name).Msg("built query")
```
Comment (ldelossa) [inline internal/vulnstore/postgres/get.go]: Is there value in displaying a hash of the query?

File: package.go [hunk 4]
```diff
@@ -4,17 +4,17 @@ type Package struct {
...
-	// the version of the distribution
+	// the version of the package
 	Version string `json:"version"`
 	// type of package. currently expectations are binary or source
-	Kind string `json:"kind"`
+	Kind string `json:"kind,omitempty"`
 	// if type is a binary package a source package maybe present which built this binary package.
 	// must be a pointer to support recursive type:
-	Source *Package `json:"source"`
+	Source *Package `json:"source,omitempty"`
```
Comment (ldelossa) [inline package.go]: omit empty can break downstream clients no? I usually avoid the need to explain "this field might be there" to ppl utilizing structs via json.

File: rpm/matcher.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+// Matcher implements driver.Matcher.
+type Matcher struct{}
+
+var _ driver.Matcher = (*Matcher)(nil)
+
+// Filter implements driver.Matcher.
+func (*Matcher) Filter(record *claircore.IndexRecord) bool {
+	// record.Repository.Kind != "rpm"
```
Comment (ldelossa) [inline rpm/matcher.go]: This is wrong in the case when the record stream this Matcher is dealing with has more then just RPM records. This is an issue of having a single rpm `Matcher`. You must still filter for RPM records or you'll attempt to query non-rpm packages.

File: rpm/packagescanner.go [hunk 6]
```diff
@@ -129,7 +132,7 @@ func (ps *Scanner) ScanContext(ctx context.Context, layer *claircore.Layer) ([]*
 	found := make([]string, 0)
 	for k, score := range possible {
 		if score == len(dbnames) {
-			found = append(found, k)
+			found = append(found, filepath.Join("/", k))
```
Comment (ldelossa) [inline rpm/packagescanner.go]: Isn't this the case for **just** RHEL containers? Not other RPM based distros?

File: osrelease/scanner.go [hunk 7]
```diff
@@ -123,21 +139,48 @@ func parse(ctx context.Context, r io.Reader) (*claircore.Distribution, error) {
 
 		switch key {
 		case "ID":
+			log.Debug().Msg("found ID")
```
Comment (ldelossa) [inline osrelease/scanner.go]: I think these debug messages should go in favor or a single one that summarizes the fully realized dist struct.

File: osrelease/scanner.go [hunk 8]
```diff
@@ -123,21 +139,48 @@ func parse(ctx context.Context, r io.Reader) (*claircore.Distribution, error) {
...
+			log.Debug().Msg("found VERSION")
 			d.Version = value
 		case "ID_LIKE":
 		case "VERSION_CODENAME":
+			log.Debug().Msg("found VERISON_CODENAME")
 			d.VersionCodeName = value
 		case "PRETTY_NAME":
+			log.Debug().Msg("found PRETTY_NAME")
+			d.PrettyName = value
+		case "REDHAT_BUGZILLA_PRODUCT":
```
Comment (ldelossa) [inline osrelease/scanner.go]: I would feel better if this was done correctly and we plumbed it into the datamodel. If it's a field that's worth having on the os-release file we don't need to munge it.

Comment (ldelossa) [reply osrelease/scanner.go]: Lets void this for now. I think this becomes a moot point once we introduce inverted index.

File: internal/vulnstore/postgres/get.go (hunk 3 above)
Comment (hdonnay) [reply internal/vulnstore/postgres/get.go]: I put it in there as assurance that a query is changing with a different set of matchexps provided

File: rpm/matcher.go (hunk 5 above)
Comment (hdonnay) [reply rpm/matcher.go]: There's no way to filter for rpm records

File: osrelease/scanner.go (hunk 8 above)
Comment (hdonnay) [reply osrelease/scanner.go]: you mean by loading the os-release file into something like an hstore? That's currently my thinking on how we should approach this.

File: rpm/packagescanner.go (hunk 6 above)
Comment (hdonnay) [reply rpm/packagescanner.go]: No, this makes sure the paths are absolute.

File: rpm/matcher.go (hunk 5 above)
Comment (ldelossa) [reply rpm/matcher.go]: @hdonnay I see.
I would create per-distro matchers in this case like the others. So a RHEL matcher which filters for something in record.Dist.

File: osrelease/scanner.go (hunk 8 above)
Comment (ldelossa) [reply osrelease/scanner.go]: @hdonnay right now the inverted_index branch has us just injesting the vulnerabilities as jsonb documents to obtain the same thing as Hstore in principal. You can make a GIN index on hstore also. But I implemented as jsonb so far.

File: package.go (hunk 4 above)
Comment (hdonnay) [reply package.go]: this can already be there or not, because it's a pointer. this just makes the resulting json smaller.

Comment (ldelossa) [reply package.go]: @hdonnay the pointer is used to support a recursive datastructure - not so much to support a nil value. 

I can see the point, but won't this make clients more brittle ? And our openapi docs less reliable? I guess right now we return an empty value for it.

**Pull request merged**